### PR TITLE
Add separated TSP specs for currently supported APIs and proposed APIs

### LINKED ISSUE #61: Provide 2 TSP specifications for currently supported APIs and proposed APIs
Right now the TSP specification contains currently supported and used APIs as well as proposed APIs.
- It's not clear what is actually used and supported and what is a proposal.
- Split the TSP yaml description into 2 files:
> 1. Currently implemented and supported APIs
> 2. Future, proposed APIs (+ APIs of bullet 1)
- Check if https://github.com/swagger-api/swagger-core can be used to automatically generate 1's.

### PR DESCRIPTION
**Pull request by bhufmann**
Fixes #61

Signed-off-by: Bernd Hufmann <Bernd.Hufmann@ericsson.com>

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -26,39 +42,31 @@ pip install -r requirements.txt
...
+[1]: https://projects.eclipse.org/projects/tools.tracecompass/developer
+[2]: https://projects.eclipse.org/projects/tools.tracecompass.incubator/developer
+[3]: http://localhost:8080/tsp/api/openapi.yaml
+[4]: https://github.com/swagger-api/swagger-core/wiki/Swagger-2.X---Integration-and-configuration#openapiresource
+[5]: https://theia-ide.github.io/trace-server-protocol/
+[6]: https://theia-ide.github.io/trace-server-protocol/swagger/
+[7]: https://marketplace.visualstudio.com/items?itemName=42Crunch.vscode-openapi
+[tspClient]: https://github.com/theia-ide/tsp-typescript-client
+[tcServer]: https://download.eclipse.org/tracecompass.incubator/trace-server/rcp/
+[apiProposed]: https://theia-ide.github.io/trace-server-protocol/proposed/
```
Comment (bhufmann) [inline README.md:72]: This will be added in a separate PR, because it will needs to be done on the `gh-pages` branch.

Comment (marco-miller) [reply README.md:72]: Remove `openapi.yaml`, now that it is being replaced with `API.yaml`?

Comment (marco-miller) [reply README.md:72]: Also, these (many) diffs look like being more than just the act of splitting the yaml files or APIs. I find this sole commit hard to review. Can we split these into more than just one commit, so we can review by iterating through each scoped change? The commits can stay in this same PR no problem I think. -Thanks.

Comment (bhufmann) [reply README.md:72]: > Remove `openapi.yaml`, now that it is being replaced with `API.yaml`?

Makes sense!

Comment (bhufmann) [reply README.md:72]: > Also, these (many) diffs look like being more than just the act of splitting the yaml files or APIs. I find this sole commit hard to review. Can we split these into more than just one commit, so we can review by iterating through each scoped change? The commits can stay in this same PR no problem I think. -Thanks.

Well, the diffs are mainly because the original references were numbered and I had remove some and re-number the references. Beside different numbers, the text is pretty much the same. Unfortunately, github diff shows to many changes which are not really changes.  I can split it multiple commits, but I don't really see it's worth because the changes are due to the update.

**Comment by bhufmann** [comment]
@marco-miller , with this patch the `API.yaml` contains the generated API from the Trace Compass trace-server. However, while having the correct endpoints, data structures have different names than what is actually serialized in this server, as well as is deserialized in the [tsp-typescript-client](https://github.com/theia-ide/tsp-typescript-client) and the [tsp-python-client](https://github.com/theia-ide/tsp-python-client). So, I'm not fully comfortable merging the `API.yaml` as proposed in this PR. This will cause confusion and it will make the client implementation fail if someone implements a different server and uses the specified names.

So, as a first option we could update the implementation in the serialization/deserialization to match the generated API. This will have huge impact in the Trace Compass server implementation. All the clients have to be updated as well.

As second option, manually modify the generated API.yaml to match the implementation before merging. This is easier to do, but error-prone due the manual updates.

A third option is to match the names of the generated names with the names of the server implementation. It has impact in the server api generation and needs to be tried out.

I would go for option 3 if it's works, or option 2. Option one is to much effort to implement.

Please let me know what you think.

**Comment by marco-miller** [comment]
> @marco-miller , with this patch the `API.yaml` contains the generated API from the Trace Compass trace-server. However, while having the correct endpoints, data structures have different names than what is actually serialized in this server, as well as is deserialized in the [tsp-typescript-client](https://github.com/theia-ide/tsp-typescript-client) and the [tsp-python-client](https://github.com/theia-ide/tsp-python-client).

Yes; we knew about this already, as the new names were our decision made during Q4 2021. (We had no other feasible choice back then.)

> So, I'm not fully comfortable merging the `API.yaml` as proposed in this PR. This will cause confusion and it will make the client implementation fail if someone implements a different server and uses the specified names.

I think we should stop manually maintaining the legacy API.yaml no matter what; it became inaccurate anyway over time.

> So, as a first option we could update the implementation in the serialization/deserialization to match the generated API. This will have huge impact in the Trace Compass server implementation. All the clients have to be updated as well.

Yes, and we did postpone that option for much later back then.

> As second option, manually modify the generated API.yaml to match the implementation before merging. This is easier to do, but error-prone due the manual updates.

I'd automate this through python scripting, so would favour this option if not manual.

> A third option is to match the names of the generated names with the names of the server implementation. It has impact in the server api generation and needs to be tried out.

I don't understand this 3rd option, or what is meant by "match"; maybe clarify a bit? Thanks.

**Comment by bhufmann** [comment]
Alright, did some more investigation and I came to some realization. The structures in the `API.yaml` are schemas and the name of the schema are not part of the JSON string in transported. For example the schema for a trace is called `ITrace` in the currently generated API.yaml file with swagger:
image

The corresponding JSON (example) is:
image

The important part here is that the fields, for example `nbEvents` or `UUID`, are matching.

I think we can go ahead and use the the generated API by Trace Compass swagger as the current specification in `API.yaml`. 

Having said that, I propose to change the generated data structures and remove the prefix `I` from it. Note the prefix `I` was added because in the Trace Compass server, interfaces where used for the API generation and the naming convention for the Trace Compass incubator project is to have such prefix for an interface. I think, removing the prefix will remove some confusion down the road for the API documentation. Besides, the swagger endpoint of the running server will provide the exact same API.yaml file.

So, I'll update this PR to remove the prefix `I` and provide a patch in the Trace Compass server to do the same for the swagger part.

**Comment by marco-miller** [comment]
Thanks @bhufmann for these latest updates. I ran out of time for reviews today but will get back to this tomorrow.

File: README.md [hunk 2]
```diff
@@ -45,9 +45,8 @@ In the meantime, Swagger has recently been added to trace-server.
 * Such reference and future (forecasted) TSP versions will be documented accordingly as these yaml files evolve.
 
 1. Import all [TC][tracecompass] and [incubator][incubator] projects in Eclipse; branch, Target Platform and API Baseline set to `master`.
-1. Run trace-server or project [1] below as Eclipse Application configured this way (or similar):
-   * `-Xms512m -Xmx1g`.
-   * Launch with `all workspace and enabled target plug-ins`.
+1. Open `traceserver.product` file in plug-in `org.eclipse.tracecompass.incubator.trace.server.product`
```
Comment (marco-miller) [inline README.md:52]: 1. Missing a trailing dot here and below, consistently with other listed items around.
2. Do we still need that reference [1] then? (Would remove it here too if not.)

File: README.md [hunk 3]
```diff
@@ -8,12 +8,26 @@ The protocol is meant to be RESTful, over HTTP.
 
 The specification is currently written in **OpenAPI 3.0** and can be pretty-visualized in the [github pages][tspGhPages].
 
-## HOW-TO
+## Current version
+
+The current version of specification is currently implemented and supportd in the [Trace Compass trace-server][tcServer] (reference implementation) and what is currently supported by the [tsp-typescript-client][tspClient].
```
Comment (marco-miller) [inline README.md:13]: supported

File: README.md [hunk 4]
```diff
@@ -8,12 +8,26 @@ The protocol is meant to be RESTful, over HTTP.
...
+
+Swagger can be used to generate the API version implemented in Trace Compass trace-server (see [here](#gernerate-the-specification)).
+
+## Future version
+
+Some proposal for additional endpoints and features are documented in the `./API-proposed.yaml`. All the proposed changes are still not confirmed and can change. The pretty-visualized file can be found [here][apiProposed]. A diff of the current version and future version will show the differences.
+
+Once a update has been approved it will be migrated to the main `./API.yaml` file.
```
Comment (marco-miller) [inline README.md:21]: Once an update

File: README.md (hunk 3 above)
Comment (marco-miller) [inline README.md:13]: of the specification (Odd read otherwise I find.)

File: README.md [hunk 5]
```diff
@@ -8,12 +8,26 @@ The protocol is meant to be RESTful, over HTTP.
...
-## HOW-TO
+## Current version
+
+The current version of specification is currently implemented and supportd in the [Trace Compass trace-server][tcServer] (reference implementation) and what is currently supported by the [tsp-typescript-client][tspClient].
+
+Swagger can be used to generate the API version implemented in Trace Compass trace-server (see [here](#gernerate-the-specification)).
+
+## Future version
+
+Some proposal for additional endpoints and features are documented in the `./API-proposed.yaml`. All the proposed changes are still not confirmed and can change. The pretty-visualized file can be found [here][apiProposed]. A diff of the current version and future version will show the differences.
```
Comment (marco-miller) [inline README.md:19]: We should note also that such diffs will also show the ordering, which differs between swagger-core's and this manual yaml. Ordering also even differs between different runs of trace-server and swagger-core init.

File: README.md [hunk 6]
```diff
@@ -8,12 +8,26 @@ The protocol is meant to be RESTful, over HTTP.
...
+
+Once a update has been approved it will be migrated to the main `./API.yaml` file.
+
+## Update manually
 
 The specification should be edited with the [OpenAPI (Swagger) Editor extension][vscodeOpenapi] for VS Code.
 
 The latter extension is assumed for consistent formatting of the `./API.yaml` file over time.
```
Comment (marco-miller) [inline README.md:27]: I'd rather not open the door to manually updating API.yaml, which is now generated from swagger-core. Rename here to ./API-proposed.yaml?

File: README.md [hunk 7]
```diff
@@ -26,38 +40,28 @@ pip install -r requirements.txt
...
+* Below is how to generate the TSP version, according to Swagger in trace-server.
+* The generated TSP should match the current supported version of the TSP. Any differences may be pushed for review.
+* `API.yaml` shows now the current supported version of the TSP.
 
 1. Import all [TC][tracecompass] and [incubator][incubator] projects in Eclipse; branch, Target Platform and API Baseline set to `master`.
 1. Open `traceserver.product` file in plug-in `org.eclipse.tracecompass.incubator.trace.server.product`
 1. Click on the `Run` button on the top right corner of the opened `traceserver.product`
 1. Browse [to here][apiyaml] ([swagger][swagger]) or so to generate server's TSP.
-1. Bring the resulting file over; e.g.: `mv ~/Downloads/openapi.yaml .`
+1. The resulting is stored in the user's Downloads directory; e.g.: `~/Downloads/openapi.yaml`
```
Comment (marco-miller) [inline README.md:55]: The resulting (?) is

Comment (marco-miller) [inline README.md:55]: `openapi.yaml` has been renamed herein; update this text too, accordingly?

File: README.md [hunk 8]
```diff
@@ -26,38 +40,28 @@ pip install -r requirements.txt
...
 1. Update the latter with its license information and remove extra information: `./openapi.py`
```
Comment (marco-miller) [inline README.md:56]: This .py script also likely needs updates in order to verify this whole change. Maybe you had it done locally so you could test these updated steps. (Include in this commit?)

**Review by marco-miller — CHANGES_REQUESTED** [review]
Multiple comments herein; keep up and thanks.

File: README.md (hunk 5 above)
Comment (bhufmann) [reply README.md:19]: I took the effort to in this PR to have the same order and the diff will show only the additional, proposed changes. However, I re-generating the API can change orderings. I think we should make the effort to keep the diff that way in the future.

File: README.md (hunk 2 above)
Comment (bhufmann) [reply README.md:52]: > 2\. Do we still need that reference [1] then? (Would remove it here too if not.)

Yes, we do.

File: README.md [hunk 9]
```diff
@@ -26,38 +40,28 @@ pip install -r requirements.txt
...
-1. The resulting git diff may then be pushed for review, at will.
-
-[1] org.eclipse.tracecompass.incubator.trace.server.jersey.rest.core
```
[the remark this answers is not in the document]
Comment (bhufmann) [reply README.md:56]: Ok. I'll keep it in this commit.

File: README.md (hunk 5 above)
Comment (bhufmann) [reply README.md:19]: I add a comment below the instructions about generating the API.

File: README.md (hunk 8 above)
Comment (marco-miller) [reply README.md:57]: This PR, AFAIK, aims at renaming openapi.yaml to API.yaml, and the previous API.yaml to API-proposed.yaml. Hence openapi.py requiring a corresponding update too. (Right?)

File: README.md (hunk 2 above)
Comment (marco-miller) [reply README.md]: > > 2. Do we still need that reference [1] then? (Would remove it here too if not.)
> 
> Yes, we do.

We ended up not needing it anymore ([1]), and you did remove it in the sibling commit, so all fine.

File: README.md (hunk 8 above)
Comment (bhufmann) [reply README.md:57]: I added a step to copy the generated openapi.yaml file to this repo before running openapi.py. So, it still works. 
Maybe I need to be clearer which files to diff (API.yaml and openapi.yaml). Also, that the matching updates should be transferred to API-proposed.yaml) and then at the end openapi.yaml can be removed when not needed anymore.

File: README.md [hunk 10]
```diff
@@ -26,39 +40,31 @@ pip install -r requirements.txt
...
-1. Run trace-server or project [1] below as Eclipse Application configured this way (or similar):
-   * `-Xms512m -Xmx1g`.
-   * Launch with `all workspace and enabled target plug-ins`.
-1. Browse [to here][4] ([swagger][5]) or so to generate server's TSP.
-1. Bring the resulting file over; e.g.: `mv ~/Downloads/openapi.yaml .`
+1. Import all [TC][tracecompass] and [incubator][incubator] projects in Eclipse; branch, Target Platform and API Baseline set to `master`.
+1. Open `traceserver.product` file in plug-in `org.eclipse.tracecompass.incubator.trace.server.product`.
+1. Click on the `Run` button on the top right corner of the opened `traceserver.product`.
+1. Browse [to here][apiyaml] ([swagger][swagger]) or so to generate server's TSP.
+1. The resulting file is stored in the user's Downloads directory; e.g.: `~/Downloads/openapi.yaml`
```
Comment (marco-miller) [inline README.md:55]: Trailing dot missing, consistently with sibling item lines.

File: README.md [hunk 11]
```diff
@@ -26,39 +40,31 @@ pip install -r requirements.txt
...
+
+[tracecompass]: https://projects.eclipse.org/projects/tools.tracecompass/developer
+[incubator]: https://projects.eclipse.org/projects/tools.tracecompass.incubator/developer
+[apiyaml]: http://localhost:8080/tsp/api/openapi.yaml
+[swagger]: https://github.com/swagger-api/swagger-core/wiki/Swagger-2.X---Integration-and-configuration#openapiresource
+[tspGhPages]: https://theia-ide.github.io/trace-server-protocol/
+[vscodeOpenapi]: https://marketplace.visualstudio.com/items?itemName=42Crunch.vscode-openapi
+[tspClient]: https://github.com/theia-ide/tsp-typescript-client
+[tcServer]: https://download.eclipse.org/tracecompass.incubator/trace-server/rcp/
+[apiProposed]: https://theia-ide.github.io/trace-server-protocol/proposed/
```
Comment (marco-miller) [inline README.md:70]: One last top commit could be added to this PR, to sort all these references alphabetically based on their tag. This would ease manual or visual look-ups, as well as consistent maintenance of their tags.

File: README.md [hunk 12]
```diff
@@ -26,39 +40,31 @@ pip install -r requirements.txt
...
-[1] org.eclipse.tracecompass.incubator.trace.server.jersey.rest.core
-
-[2]: https://projects.eclipse.org/projects/tools.tracecompass/developer
-[3]: https://projects.eclipse.org/projects/tools.tracecompass.incubator/developer
-[4]: http://localhost:8080/tsp/api/openapi.yaml
-[5]: https://github.com/swagger-api/swagger-core/wiki/Swagger-2.X---Integration-and-configuration#openapiresource
-[6]: https://theia-ide.github.io/trace-server-protocol/
-[7]: https://theia-ide.github.io/trace-server-protocol/swagger/
-[8]: https://marketplace.visualstudio.com/items?itemName=42Crunch.vscode-openapi
+1. The resulting diff between `API.yaml` and `openapi.yaml` can then be pushed for review. Note, that the order of fields, components etc. might be different everytime the API is generated using swagger-core. This is due to how swagger-core is implemented.
```
Comment (marco-miller) [inline README.md:58]: Can we sub-itemize this rather long item sentence? This would ease readability I find.

**Review by marco-miller — CHANGES_REQUESTED** [review]
Thanks for the latest updates. Just a few minor comments left for me.

File: README.md (hunk 12 above)
Comment (bhufmann) [reply README.md:58]: Ok

File: README.md (hunk 11 above)
Comment (bhufmann) [reply README.md:71]: will do

File: README.md [hunk 13]
```diff
@@ -26,39 +40,32 @@ pip install -r requirements.txt
...
-[1] org.eclipse.tracecompass.incubator.trace.server.jersey.rest.core
-
-[2]: https://projects.eclipse.org/projects/tools.tracecompass/developer
-[3]: https://projects.eclipse.org/projects/tools.tracecompass.incubator/developer
-[4]: http://localhost:8080/tsp/api/openapi.yaml
-[5]: https://github.com/swagger-api/swagger-core/wiki/Swagger-2.X---Integration-and-configuration#openapiresource
-[6]: https://theia-ide.github.io/trace-server-protocol/
-[7]: https://theia-ide.github.io/trace-server-protocol/swagger/
-[8]: https://marketplace.visualstudio.com/items?itemName=42Crunch.vscode-openapi
+1. The resulting diff between `API.yaml` and `openapi.yaml` can then be pushed for review.
```
Comment (marco-miller) [inline README.md:58]: Trailing space, to remove; `markdownlint` tells you so in VS Code.

**Review by marco-miller — CHANGES_REQUESTED** [review]
Nit left.

File: README.md (hunk 13 above)
Comment (bhufmann) [reply README.md:58]: ok

**Review by marco-miller — APPROVED** [review]
Thanks!

**Review by PatrickTasse — APPROVED** [review]
